test/framework/multi_node_control_plane.go

```diff
@@ -0,0 +1,128 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package framework
+
+import (
+	"context"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// MultiNodeControlplaneClusterInput defines the necessary dependencies to run a multi-node control plane cluster.
+type MultiNodeControlplaneClusterInput struct {
+	Management        ManagementCluster
+	Cluster           *clusterv1.Cluster
+	InfraCluster      runtime.Object
+	ControlplaneNodes []Node
+	CreateTimeout     time.Duration
+}
+
+// SetDefaults defaults the struct fields if necessary.
+func (m *MultiNodeControlplaneClusterInput) SetDefaults() {
+	if m.CreateTimeout == 0 {
+		m.CreateTimeout = 10 * time.Minute
+	}
+}
+
+// MulitNodeControlPlaneCluster creates an n node control plane cluster.
+// Assertions:
+//  * The number of nodes in the created cluster will equal the number of nodes in the input data.
+func MultiNodeControlPlaneCluster(input *MultiNodeControlplaneClusterInput) {
+	ctx := context.Background()
+	Expect(input.Management).ToNot(BeNil())
+
+	mgmtClient, err := input.Management.GetClient()
+	Expect(err).NotTo(HaveOccurred(), "stack: %+v", err)
+
+	By("creating an InfrastructureCluster resource")
+	Expect(mgmtClient.Create(ctx, input.InfraCluster)).NotTo(HaveOccurred())
+
+	By("creating a Cluster resource linked to the InfrastructureCluster resource")
+	Expect(mgmtClient.Create(ctx, input.Cluster)).NotTo(HaveOccurred())
+
+	// Wait for the cluster infrastructure
+	Eventually(func() string {
+		cluster := &clusterv1.Cluster{}
+		key := client.ObjectKey{
+			Namespace: input.Cluster.GetNamespace(),
+			Name:      input.Cluster.GetName(),
+		}
+		if err := mgmtClient.Get(ctx, key, cluster); err != nil {
+			return err.Error()
+		}
+		return cluster.Status.Phase
+	}, input.CreateTimeout, 10*time.Second).Should(Equal(string(clusterv1.ClusterPhaseProvisioned)))
+
+	// create all the machines at once
+	for _, node := range input.ControlplaneNodes {
+		By("creating an InfrastructureMachine resource")
+		Expect(mgmtClient.Create(ctx, node.InfraMachine)).NotTo(HaveOccurred())
+
+		By("creating a bootstrap config")
+		Expect(mgmtClient.Create(ctx, node.BootstrapConfig)).NotTo(HaveOccurred())
+
+		By("creating a core Machine resource with a linked InfrastructureMachine and BootstrapConfig")
+		Expect(mgmtClient.Create(ctx, node.Machine)).NotTo(HaveOccurred())
+	}
+
+	// wait for all the machines to be running
+	By("waiting for all machines to be running")
+	for _, node := range input.ControlplaneNodes {
+		Eventually(func() string {
```

Comment:
I think this might end up being serial, not parallel. Not sure if that's intended.

Reply:
This will be serial and not parallel, it is intended.

The usual case is the first one becomes ready then the others join the cluster and become ready at about the same time. This will mostly be waiting for the first one then the second and third `Eventually`s finish more quickly